crypto/src/key/secp256k1/extended_keys.rs

```diff
@@ -0,0 +1,256 @@
+use hmac::{Hmac, Mac};
+use secp256k1;
+use secp256k1::SECP256K1;
+use serialization::{Decode, Encode};
+use sha2::Sha512;
+use zeroize::Zeroize;
+
+// Create alias for HMAC-SHA512
+type HmacSha512 = Hmac<Sha512>;
+
+#[derive(Debug, PartialEq, Eq, Clone, Encode, Decode)]
+pub struct Secp256k1ExtendedPrivateKey {
+    /// Chain code
+    pub chain_code: ChainCode,
+    /// Private key
+    pub private_key: Secp256k1PrivateKey,
+}
+
+impl Secp256k1ExtendedPrivateKey {
+    pub fn new_master(seed: &[u8]) -> Result<Secp256k1ExtendedPrivateKey, DerivationError> {
+        let mut mac =
+            HmacSha512::new_from_slice(b"Bitcoin seed").expect("HMAC can take key of any size");
+        mac.update(seed);
+        let mut result = mac.finalize().into_bytes();
+
+        let private_key = secp256k1::SecretKey::from_slice(&result[..32])
+            .map_err(|_| DerivationError::KeyDerivationError)?
+            .into();
+
+        let chain_code: [u8; CHAINCODE_LENGTH] =
+            result[32..].try_into().expect("Chaincode size is 32 bytes");
+        let chain_code = ChainCode::from(chain_code);
+
+        result.zeroize();
+
+        Ok(Secp256k1ExtendedPrivateKey {
+            private_key,
+            chain_code,
+        })
+    }
+
+    pub fn new<R: Rng + CryptoRng>(
+        rng: &mut R,
+    ) -> (Secp256k1ExtendedPrivateKey, Secp256k1ExtendedPublicKey) {
+        // Create a new chain code
+        let mut chain_code = [0u8; 32];
+        rng.fill_bytes(&mut chain_code);
+        let chain_code = chain_code.into();
+        let private_key = secp256k1::SecretKey::new(rng).into();
+        // Generate a new private key
+        let ext_priv = Secp256k1ExtendedPrivateKey {
+            private_key,
+            chain_code,
+        };
+        // Generate the public key
+        let ext_pub = Secp256k1ExtendedPublicKey::from_private_key(&ext_priv);
+        // Return the pair
+        (ext_priv, ext_pub)
+    }
+
+    pub fn as_private_key(&self) -> &Secp256k1PrivateKey {
+        &self.private_key
+    }
+}
+
+impl Derivable for Secp256k1ExtendedPrivateKey {
+    fn derive_child(self, num: ChildNumber) -> Result<Self, DerivationError> {
+        let mut mac = HmacSha512::new_from_slice(&self.chain_code.into_array())
+            .expect("HMAC can take key of any size");
+
+        let secp_key = self.private_key.data;
+
+        // We only support hard derivations
+        if num.is_hardened() {
+            mac.update(&[0u8]);
+            mac.update(&secp_key[..]);
+        } else {
+            return Err(DerivationError::UnsupportedDerivationType);
+        }
+
+        // Add the child number
+        mac.update(&num.to_encoded_be_bytes());
+
+        let mut result = mac.finalize().into_bytes();
+        let private_key = secp256k1::SecretKey::from_slice(&result[..32])
+            .map_err(|_| DerivationError::KeyDerivationError)?
+            // TODO [SECURITY] a Scalar is created here but not erased
```

Comment:
We usually follow the format TODO(TAG). It makes it easier to search for all tagged TODOs.